Add scripts and configs for hyperparameter tuning

taufeeque9 opened the pull request:
## Description

This PR adds the scripts and configs for tuning hyperparameters of imitation algorithms. The results in the first draft of the [paper](https://arxiv.org/abs/2211.11972) were computed using the configs in the PR. Note that the preference comparisons algorithm still needs to be tuned on all of the environments.

Changes:

- `imitation/scripts/parallel.py`: Added features useful for tuning hyperparameters (HPs) of the algorithms. Specifically, added options to 1) repeat trials across configs to get a better (mean) estimate of the return of HP configs, 2) evaluate the best config from an HP tune run, and 3) restart failed trials of a tune run.
- `imitation/scripts/analyze.py`: Added table verbosity level 3, which generates a CSV with all the config HPs, including RL HPs and the specific algorithm's arguments, along with the expert returns and the returns of the imitation algorithm.
- `imitation/scripts/config/parallel.py`: Added HP search space for various algorithms like BC, Dagger, GAIL, and AIRL. Search space for Preference Comparisons may not be good enough.
- `imitation/scripts/config/train_*.py`: The added configs contain the HP for the tuned base RL algorithm.


## Testing

Updated test cases for `parallel` and tested HP tuning manually.

ernestum reviewed (changes requested):
First round of review mostly with suggestions to the parallel script.

ernestum commented:
Not part of this PR but @taufeeque9  could you elaborate on the `benchmarking/util.py` script? It is unclear right now where in the benchmark pipeline it is used.

ernestum commented on src/imitation/scripts/config/train_rl.py:
```diff
@@ -70,8 +72,30 @@ def cartpole():
-    total_timesteps = int(1e6)
+    environment = dict(gym_id="seals/CartPole-v0", num_vec=8)
+    total_timesteps = int(1e5)
+    policy = dict(
+        policy_cls="MlpPolicy",
+        policy_kwargs=dict(
+            activation_fn=nn.ReLU,
+            net_arch=[dict(pi=[64, 64], vf=[64, 64])],
+        ),
+    )
+    normalize_reward = False
+    rl = dict(
+        batch_size=4096,
+        rl_kwargs=dict(
+            batch_size=256,
+            clip_range=0.4,
+            ent_coef=0.008508727919228772,
+            gae_lambda=0.9,
+            gamma=0.9999,
+            learning_rate=0.0012403278189645594,
+            max_grad_norm=0.8,
+            n_epochs=10,
+            vf_coef=0.489343896591493,
+        ),
+    )
```
I think we should somehow document where all those new (also below and in other config files) hyperparameter values come from. Otherwise, it is unclear why we should trust them. What script generated them? What was the HP search space? How large the optimization budget?

ernestum commented on src/imitation/scripts/train_adversarial.py:
```diff
@@ -167,6 +167,7 @@ def callback(round_num: int, /) -> None:
     return {
         "imit_stats": imit_stats,
         "expert_stats": rollout.rollout_stats(expert_trajs),
+        "mean_return": imit_stats["monitor_return_mean"],
```
Unclear why you set `mean_return` to `monitor_return_mean` here. Can we at least document this quirk of the `eval_policy` output? We also seem to do it in `train_rl.py` and `train_preference_comparisons.py`. So maybe add a function that just does this weird copying and document the reason in it's docstring to make this explicit.

taufeeque9 replied:
Removed the need to do all of this. Earlier, I used this hack to have a single location in the returned dict for the metric to be specified in `ray.tune.run`. Simplified this by using the original location of the metric for the different algorithms.

ernestum commented on src/imitation/scripts/config/train_adversarial.py:
```diff
@@ -242,7 +239,6 @@ def seals_swimmer():
 
 @train_adversarial_ex.named_config
 def seals_walker():
-    locals().update(**MUJOCO_SHARED_LOCALS)
```
Careful! While the others were commented out, this one was not. Are you sure you want to remove this line?

taufeeque9 replied:
Yes, this line was vestigial as the `MUJOCO_SHARED_LOCALS` are anyways getting updated with the config below it.

ernestum commented on src/imitation/scripts/parallel.py:
```diff
@@ -95,15 +126,110 @@ def parallel(
+                # update cpus per trial only if it is provided in `resources_per_trial`
+                # Uses the default values (cpu=1) if it is not provided
+                if "cpu" in resources_per_trial:
+                    resources_per_trial_eval = copy.deepcopy(resources_per_trial)
```
What is the point of this variable? It is not used anywhere.

ernestum commented on src/imitation/scripts/analyze.py:
```diff
@@ -262,26 +265,35 @@ def analyze_imitation(
         The DataFrame generated from the Sacred logs.
     """
-    table_entry_fns_subset = _get_table_entry_fns_subset(table_verbosity)
+    if table_verbosity == 3:
```
Could you add a note to the code to explain this edge case?

ernestum commented on src/imitation/scripts/analyze.py:
```diff
@@ -262,26 +265,35 @@ def analyze_imitation(
+        table_entry_fns_subset = _get_table_entry_fns_subset(table_verbosity)
 
-    rows = []
+    df = pd.DataFrame()
```
Can we have a bit more expressive variable names than just `df` and `new_df`. What does the df contain? What is the purpose of `new_df`?

ernestum commented on src/imitation/scripts/parallel.py:
```diff
@@ -95,15 +126,114 @@ def parallel(
+            key_prefix = "imit_stats/"
+        key = key_prefix + "monitor_return_mean"
+
+        if eval_best_trial:
```
After looking at this again I would have a strong preference to exclude the result evaluation from the `parallel.py` script. This scrip is already way to complex and grown over the years. Now we add another feature to it and I fear it will become entirely incomprehensible in the future. I am happy to discuss other solutions.

taufeeque9 commented:
> Not part of this PR but @taufeeque9 could you elaborate on the `benchmarking/util.py` script? It is unclear right now where in the benchmark pipeline it is used.

Running the HP tuning script using `parallel.py` generates JSON files for all the HP tuning config that contains the sacred config that can be used to re-run the experiment with the exact arguments. 

The `benchmarking/util.py` file contains the code to clean the generated JSON files by removing specific arguments like seed, rollout path, etc., that were specific for the run. The cleaned files are what we have uploaded as `benchmarking/*.json` for all the algorithms and the environments.

ernestum commented:
Just to make sure that I understand it correctly:

For each algorithm in {BC, AIRL, GAIL, DAgger} and each environment in {ant, half_cheetah, hopper, swimmer, walker} we tune the hyperparameters using `parallel.py`.
This means we have 4*5 = 20 runs of `parrallel.py`.

This generates some amount of files.

From those files we extract the best hyperparameter configurations using `benchmarking/utils.py` and place them in the json files in `benchmarking/`.

To reproduce a run with one of those runs with the tuned hyperparameters we run `python -m imitation.scripts.<train_script> <algo> with benchmarking/<config_name>.json`

taufeeque9 commented:
> Just to make sure that I understand it correctly:
> 
> For each algorithm in {BC, AIRL, GAIL, DAgger} and each environment in {ant, half_cheetah, hopper, swimmer, walker} we tune the hyperparameters using `parallel.py`. This means we have 4*5 = 20 runs of `parrallel.py`.
> 
> This generates some amount of files.
> 
> From those files we extract the best hyperparameter configurations using `benchmarking/utils.py` and place them in the json files in `benchmarking/`.
> 
> To reproduce a run with one of those runs with the tuned hyperparameters we run `python -m imitation.scripts.<train_script> <algo> with benchmarking/<config_name>.json`.
> 
> Can you verify this @taufeeque9 ?

Yes, that sounds correct. The only thing to add is that we get the best hyperparameter configurations from running `parallel.py` with `eval_best_trial=True` and we use `benchmarking/utils.py` to clean the best configs that can be uploaded to the `benchmarking/` directory on GitHub.

ernestum commented on src/imitation/scripts/parallel.py:
```diff
@@ -31,14 +29,12 @@ def parallel(
-    resume: Union[str, bool] = False,
-) -> None:
+    repeat: int,
+    search_alg: Optional[str],
```
I would prefer if you did not introduce any tuning-specific arguments to the `parallel` script. One way to solve this is to add a `tune_run_kwargs` parameter and use that to inject the optuna search algo etc.

Then it becomes the responsibility of the `tuning.py` script to set everything up correctly.

taufeeque9 replied:
I think `repeat` is a good argument to keep in the `parallel` script because it allows a sampled hyperparameter config to run multiple times, which is a good general feature to have in the `parallel` script. So if we want to keep it in the `parallel` script, we must keep `search_alg` in the `parallel` script because `repeat` requires a search algorithm to be specified first.

If we don't think `repeat` is beneficial to keep in the `parallel` script, I can introduce a `tune_run_kwargs` in `parallel` and move the repeat and search algorithm options to the `tuning` script only.

ernestum replied:
I agree with the `repeat` argument. However, if we just repeat the same configuration a couple of times, why do we need a search algorithm? Maybe the Ray API needs one up just any search algorithm will do right?

taufeeque9 replied:
> if we just repeat the same configuration a couple of times, why do we need a search algorithm? 

The Ray API for repeating takes a search algorithm and wraps a repeater on top of it to give the modified search algorithm: `search.Repeater(optuna.OptunaSearch(), repeat=repeat)`.

The repeater allows running multiple trials with the same configuration sampled by the algorithm. Hence, it needs the search algorithm as an argument. I don't think we can have the repeat argument without the search algorithm.

ernestum replied:
What I meant was: we could just set the search algo to any simple search algo as soon as `repeat > 0` then we do not have to expose this quirk of the Ray API to our users.

taufeeque9 replied:
That makes sense. I have redacted `search_alg` from the arguments of the parallel script, along with some other rarely used arguments to introduce `tune_run_kwargs`. Let me know if this looks good.

ernestum reviewed (changes requested):
Thanks for the updates @taufeeque9 !
Could you also make sure to add an overview documentation with some example calls for the `tuning.py` script?

Also some rough tests for it would be nice.

taufeeque9 commented on benchmarking/tuning_config.py:
```diff
@@ -0,0 +1,237 @@
+            "demonstrations": {"source": "huggingface"},
+        },
+        search_space={
+            "config_updates": {
```
I think it is more straightforward and clear to keep the "config_updates" hierarchy as it matches the way we use config_updates in the parallel script and the tuning config.

ernestum replied:
All right.

taufeeque9 commented on benchmarking/tuning.py:
```diff
@@ -0,0 +1,102 @@
+        else:
+            return_key = "imit_stats/monitor_return_mean"
+        df = result.results_df
+        df = df[df["config/named_configs"].notna()]
```
Thanks for pointing this out! This is an unnecessary line that was useful at some point when I was benchmarking the algorithms but is not required now. I have removed it.

ernestum commented on src/imitation/scripts/parallel.py:
```diff
@@ -72,19 +72,22 @@ def parallel(
         local_dir: `local_dir` argument to `ray.tune.run()`.
         upload_dir: `upload_dir` argument to `ray.tune.run()`.
-        search_alg: can be either "optuna" or None.
+        search_alg: can be either "optuna" or None. Setting `None` allows for
```
Now it is a lot clearer what this parameter does. It is still confusing why it behaves that way. Again, I agree that we should keep the `repeat` argument but I think there should be a way to go without the `search_alg` argument.

taufeeque9 replied:
See the comment above: https://github.com/HumanCompatibleAI/imitation/pull/675#discussion_r1266249803

ernestum commented on src/imitation/scripts/parallel.py:
```diff
@@ -72,19 +72,22 @@ def parallel(
     Returns:
-        The result of `ray.tune.run()`.
+        The result of running the parallel experiment with `ray.tune.run()`.
+        Useful for fetching the configs and results dataframe of all the trials.
```
Perfect! This is the context that was missing. :+1:

ernestum commented on tests/test_benchmarking.py:
```diff
@@ -44,3 +46,28 @@ def test_benchmarks_print_config_succeeds(algorithm: str, environment: str):
+    # because running the configs requires MuJoCo.
+    # Requiring MuJoCo to run the tests adds too much complexity.
+
+    # We need to use sys.executable, not just "python", on Windows as
```
I like how you explain the use of `sys.executable` here!

ernestum reviewed (commented):
Thanks for implementing the changes I requested in the last round! The code is already a lot more readable. This round only includes a couple not nits and we still need to come to a conclusion with the `search_alg` parameter.

ernestum commented on benchmarking/README.md:
````diff
@@ -17,3 +17,24 @@ python -m imitation.scripts.<train_script> <algo> with benchmarking/<config_name
+
+To tune the hyperparameters of an algorithm using the default search space provided:
+```bash
+python tuning.py with example_{algo} 'parallel_run_config.base_named_configs=["{env}"]'
````
Why is there the `example_` prefix?

taufeeque9 replied:
A vestigial naming convention left from the earlier code of parallel. Have thought about removing it for quite some time. Will remove it now.

ernestum commented on src/imitation/scripts/config/parallel.py:
```diff
@@ -7,10 +7,8 @@
-Ex: python -m imitation.scripts.parallel with example_gail \
-    'base_named_configs=["logging.wandb_logging", "seals_half_cheetah"]'
+For tuning hyperparameters of an algorithm on a given environment,
+check out the benchmarking/tuning.py script.
```
Why did you remove the example here?

taufeeque9 replied:
The example was for tuning the HPs using the parallel script which we have now moved to the tuning script along with all the named configs for tuning HPs.

ernestum reviewed (commented):
Thanks for the changes. Looking better and better now!

ernestum commented on src/imitation/scripts/config/parallel.py:
```diff
@@ -31,19 +34,11 @@ def config():
-    upload_dir = "s3://shwang-chai/private"
+    num_samples = 1  # Number of samples per grid search configuration
+    repeat = 1  # Number of times to repeat a sampled configuration
+    search_alg = "optuna"  # search algorithm to use
```
This config can go away now right?

ernestum commented on benchmarking/tuning.py:
```diff
@@ -0,0 +1,167 @@
+    Raises:
+        ValueError: If no trials are returned by the parallel run of tuning.
+    """
+    run = parallel_ex.run(config_updates=parallel_run_config)
```
Set the `tune_run_kwargs` here to ensure that the search algorithm etc. is set correctly? It will work when `repeat>1` but not all of your configs hae that.

ernestum reviewed (commented):
This looks a lot better. Still some small things to consider.

ernestum commented on benchmarking/tuning.py:
```diff
@@ -32,7 +33,15 @@ def tune(
     """
-    run = parallel_ex.run(config_updates=parallel_run_config)
+    search_alg = optuna.OptunaSearch()
+    updated_parallel_run_config = copy.deepcopy(parallel_run_config)
```
Took me quite some time to understand how you are updating the config here (ok maybe I should have a :coffee: ). But I think this is essentially the same as this?
```python
    if "tune_run_kwargs" in updated_parallel_run_config:
        updated_parallel_run_config["tune_run_kwargs"]["search_alg"] = optuna.OptunaSearch()
    else:
        updated_parallel_run_config["tune_run_kwargs"] = dict(search_alg=optuna.OptunaSearch())
```

taufeeque9 replied:
Done!

ernestum replied:
thx!

ernestum commented on tests/scripts/test_scripts.py:
```diff
@@ -882,13 +883,13 @@ def test_parallel_train_adversarial_custom_env(tmpdir):
             demonstrations=dict(path=path),
         ),
-        search_space=dict(command_name="gail"),
+        search_space=dict(command_name=tune.choice(["gail"])),
```
Why was this needed? Can we document this inline?

taufeeque9 replied:
This was needed because the search space needs to contain at least one element (so that it can be sampled) when we use the optuna search algorithm (enabled by repeat=2).

ernestum replied:
All right. Then please document this inline because somebody (probably me) will be confused by this construct when I read it again :-)

taufeeque9 replied:
Done 👍

ernestum reviewed (commented):
This time around I did a pass over the entire PR again as the previous reviews were just incremental reviews on the most recent changes.

taufeeque9 commented on benchmarking/tuning.py:
I agree. Adding the tuning script was a great suggestion, as both the tuning and the parallel scripts are now much easier to understand. Thanks for pushing for it!

taufeeque9 commented on src/imitation/scripts/parallel.py:
```diff
@@ -95,15 +113,47 @@ def parallel(
     ray.init(**init_kwargs)
+    updated_tune_run_kwargs = copy.deepcopy(tune_run_kwargs)
+    if repeat > 1:
+        if "search_alg" not in updated_tune_run_kwargs:
+            updated_tune_run_kwargs["search_alg"] = optuna.OptunaSearch()
+        try:
+            algo = updated_tune_run_kwargs["search_alg"]
+            algo = search.Repeater(algo, repeat)
+            updated_tune_run_kwargs["search_alg"] = algo
+        except AttributeError:
+            raise ValueError(
+                "repeat > 1 but search_alg is not an instance of "
+                "ray.tune.search.SearchAlgorithm",
+            )
```
Good suggestion! I often forget about the `get` method in dictionary.

> Also: why are we using Optuna if we just want to repeat some runs? This seems like overkill and a simple grid search might have less overhead?

The point of adding `repeat` is that we want to perform multiple runs of all the sampled trials. So let's say if `learning_rate` is sampled from `loguniform(1e-5, 1e-2)`, then if a value of $1.3 * 10^{-4}$ was sampled for a trial, then our `repeat=n` argument will perform this trial with the value of $1.3 * 10^{-4}$ `n` times. I believe this cannot be done using grid search unless I'm mistaken.

Regarding Optuna, I agree that we don't have to use Optuna for repeat some runs and can instead use any other search algorithm like `tune.search.basic_variant.BasicVariantGenerator` which supports both random and grid searches.

ernestum replied:
Note from discussion with @taufeeque9:
We will stick with Optuna since the `BasicVariantGenerator` does not support being wrapped with a repeater.

Details here: https://github.com/ray-project/ray/issues/33677

taufeeque9 commented on experiments/commands.py:
The documentation of `commands.py` and related scripts are currently in the benchmarking docs PR #657.

taufeeque9 commented on src/imitation/scripts/analyze.py:
```diff
@@ -166,6 +166,9 @@ def _get_algo_name(sd: sacred_util.SacredDicts) -> str:
 
 def _return_summaries(sd: sacred_util.SacredDicts) -> dict:
     imit_stats = get(sd.run, "result.imit_stats")
+    if imit_stats is None:
+        # stored in rollout key for preference comparison
+        imit_stats = get(sd.run, "result.rollout")
```
I had already changed it in the preference comparisons script but didn't remove it from here. Thanks for pointing this out!

taufeeque9 commented on src/imitation/scripts/config/train_adversarial.py:
Those are just the hyperparameters of the tuned PPO algorithm on the respective environment. We set those as the base hyperparameters of any algorithm that uses RL as a subroutine (including preference comparisons).

I have added a note for it at the top of the script.

taufeeque9 commented on src/imitation/scripts/config/train_preference_comparisons.py:
I've added a note at the top of the script that the RL algorithm's hyperparameters were tuned using the RL Zoo library. I don't remember specific details like the tuning budget or the number of trials. Let me know if the note looks sufficient or if there are things we should definitely try to add.

taufeeque9 commented on src/imitation/scripts/parallel.py:
```diff
@@ -148,7 +198,7 @@ def _ray_tune_sacred_wrapper(
     """
 
-    def inner(config: Mapping[str, Any], reporter) -> Mapping[str, Any]:
+    def inner(config: Dict[str, Any], reporter) -> Mapping[str, Any]:
```
Good call!

AdamGleave commented on benchmarking/tuning.py:
```diff
@@ -0,0 +1,174 @@
+from pandas.api import types as pd_types
+from ray.tune.search import optuna
+from sacred.observers import FileStorageObserver
+from tuning_config import parallel_ex, tuning_ex
```
PEP8: standard library, third-party then first party-imports with line separating each.
```suggestion

from tuning_config import parallel_ex, tuning_ex
```

AdamGleave commented on benchmarking/tuning.py:
```diff
@@ -0,0 +1,174 @@
+    """
+    updated_parallel_run_config = copy.deepcopy(parallel_run_config)
+    search_alg = optuna.OptunaSearch()
+    if "tune_run_kwargs" in updated_parallel_run_config:
```
Is this equivalent?
```suggestion
tune_run_kwargs = updated_parallel_run_config.setdefault("tune_run_kwargs", dict())
tune_run_kwargs["search_alg"] = search_alg
```

taufeeque9 replied:
Yes, I wasn't aware of `setdefault`. Nice little feature!

AdamGleave commented on benchmarking/tuning.py:
```diff
@@ -0,0 +1,174 @@
+    eval_best_trial_resource_multiplier: int = 1,
+    num_eval_seeds: int = 5,
+) -> None:
+    """Tune hyperparameters of imitation algorithms using parallel script.
```
I didn't understand the separate training and evaluation stage until I read the code, consider adding some explanation of that?

AdamGleave reviewed (commented):
Are the scripts in `benchmarking/` being type checked and linted? I don't see them in the output at https://app.circleci.com/pipelines/github/HumanCompatibleAI/imitation/3897/workflows/02ec5434-d883-4edf-9fa5-99faeb0645f2/jobs/15772 The directory `benchmarking/` is not specified in the `inputs` config for ptype in `setup.cfg`. 

I think should either move the scripts into `src/imitation/scripts/` and make them part of the package, or update CI config to test things in `benchmarking/` (worth checking we're not missing them for flake8 etc as well).

AdamGleave commented on benchmarking/tuning_config.py:
```diff
@@ -0,0 +1,232 @@
+def gail():
+    parallel_run_config = dict(
+        sacred_ex_name="train_adversarial",
+        run_name="gail_tuning_hc",
```
What does "hc" stand for? Why is this only present for gail?

taufeeque9 replied:
Fixed. It got there from another run I was doing but shouldn't be there.

AdamGleave commented on src/imitation/scripts/analyze.py:
```diff
@@ -262,38 +262,50 @@ def analyze_imitation(
+        #  config columns.
+        table_entry_fns_subset = _get_table_entry_fns_subset(2)
+    else:
+        table_entry_fns_subset = _get_table_entry_fns_subset(table_verbosity)
```
Could drop if statement and write:
```suggestion
    table_entry_fns_subset = _get_table_entry_fns_subset(min(table_verbosity, 2))
```

AdamGleave reviewed (commented):
Right now it seems like there's a lot of duplication in specifying the pre-tuned hyperparameters in `*.json` files in `benchmarking/`, and in `named_configs` in each algorithm. If possible it seems preferable to make this more DRY (Don't Repeat Yourself) so that we have a single source of truth: otherwise I could imagine them drifting apart as we do future hyperparameter sweeps.

It is convenient to be able to write e.g. `with seals_ant` rather than `with benchmarking/airl_seals_ant_best_hp_eval.json`. But Sacred does support an `ex.add_named_config`: https://github.com/IDSIA/sacred/blob/17c530660d5b405af0f5c286b1a93f3d8911d026/sacred/ingredient.py#L251 So we could replace some of these with that.

It does seem like the named configs have been edited to be more minimal. That's great, but as a user, which one should I use? If the answer is "always use the named_config" then we should really consider editing the JSONs (manually or automatically). If the answer is "it depends" then it would help to document the pros and cons and justify why we need both.

AdamGleave commented on src/imitation/scripts/parallel.py:
```diff
@@ -17,29 +20,33 @@
+    to `upload_dir` if that argument is provided in `tune_run_kwargs`.
 
     Args:
         sacred_ex_name: The Sacred experiment to tune. Either "train_rl" or
```
```suggestion
        sacred_ex_name: The Sacred experiment to tune. Either "train_rl",
```

AdamGleave commented on src/imitation/scripts/parallel.py:
```diff
@@ -17,29 +20,33 @@
     Args:
         sacred_ex_name: The Sacred experiment to tune. Either "train_rl" or
-            "train_adversarial".
+            "train_imitation" or "train_adversarial" or "train_preference_comparisons".
```
```suggestion
            "train_imitation", "train_adversarial" or "train_preference_comparisons".
```

AdamGleave commented on src/imitation/scripts/parallel.py:
```diff
@@ -95,15 +113,45 @@ def parallel(
+        if experiment_checkpoint_path:
+            # load experiment analysis results
+            result = ray.tune.ExperimentAnalysis(experiment_checkpoint_path)
+            result._load_checkpoints_from_latest(
```
Unfortunate we have to use a private method to load this -- I assume there's no alternative via public methods?

taufeeque9 replied:
The `ExperimentAnalysis` class was recently rewritten in `ray.tune` and we no longer need to use the private methods. It's good that you pointed this out!

AdamGleave commented on tests/scripts/test_scripts.py:
```diff
@@ -977,7 +981,7 @@ def test_analyze_imitation(tmpdir: str, run_names: List[str], run_sacred_fn):
 
     # Check that analyze script finds the correct number of logs.
-    def check(run_name: Optional[str], count: int) -> None:
+    def check(run_name: Optional[str], count: int, table_verbosity=1) -> None:
```
```suggestion
    def check(run_name: Optional[str], count: int, table_verbosity: int = 1) -> None:
```

taufeeque9 commented on benchmarking/tuning.py:
```diff
@@ -0,0 +1,174 @@
+    grp_keys = [c for c in df.columns if c.startswith("config") and "seed" not in c]
+    grps = df.groupby(grp_keys)
+    # store mean return of runs across all seeds in a group
+    df["mean_return"] = grps[return_key].transform(lambda x: x.mean())
```
These return different number of rows. Your suggestion obtains mean return for every group whereas the transform method obtains mean return for every trial which is compatible with the number of rows in `df`. I have explained this in the comments now.

taufeeque9 commented on benchmarking/tuning.py:
```diff
@@ -0,0 +1,174 @@
+    ][0]
+
+    if print_return:
+        all_returns = df[df["mean_return"] == row["mean_return"]][return_key]
```
In case we get multiple distinct hyperparameter groups that get the same mean returns across seeds, we pick the first hyperparameter group from the `df`. And the ordering of hyperparameter groups in the `df` should be arbitrary, so it should be fine to do this, right?

taufeeque9 commented:
> I think should either move the scripts into src/imitation/scripts/ and make them part of the package

I have moved the benchmarking scripts into `src/imitation/scripts` and the tuned hyperparameter json files into `src/imitation/scripts/config/tuned_hps`. This seems more clean and elegant to me than keeping everything in the `benchmarking` directory. With these changes the benchmarking directory seems useless now as it just contains a `util.py` file to clean up the json files. We can move the `util.py` file somewhere else and delete the benchmarking directory.

> Right now it seems like there's a lot of duplication in specifying the pre-tuned hyperparameters in *.json files in benchmarking/, and in named_configs in each algorithm.

This was a great point! Thanks for noting it. I have added all of the json files as named configs in their respective `train_<algo>` script. So now we can directly include the named configs from the command line instead of the json files. As a result, I have also removed the old named configs for the algorithms & environments we have tuned.

So for example, in the `train_adversarial` script, I have removed `seals_half_cheetah` named config and added `airl_seals_half_cheetah` and `gail_seals_half_cheetah` configs for the respective algorithms of `airl` and `gail`. So now a user can run the following:
```
python -m imitation.scripts.train_adversarial airl with airl_seals_half_cheetah
```

Note that there's a redundancy in the command where `airl` is being mentioned twice. I've tried to think of ways to remove this redundancy but couldn't figure a way out that preserves the command name `airl` and removes `airl` from the named config. We can maybe remove the command name altogether and keep `algo_env` config names but that would be quite a big change for this PR. I think for now raising this as an issue might be best so that we can deal with this later with some other PR.

AdamGleave reviewed (approved):
LGTM, some minor suggestions.

CI is failing right now -- we should get it green before merging. 

Lint error is just whitespace should be easy to fix.

Unit test error is in test_scripts for eval_policy. Not sure why this is being triggered here as we've not changed the code. It may just be a Gymnasium upgrade triggering it? If so, happy for you to do a hotfix for this in a separate PR, we can then get that hotfix merged then finally merge this PR. Can probably be resolved by changing `gym.make` to include appropriate `render_mode="rgb_array"`. 

> With these changes the benchmarking directory seems useless now as it just contains a util.py file to clean up the json files. We can move the util.py file somewhere else and delete the benchmarking directory.

Agreed probably best to move `util.py` somewhere else. There's not an obvious alternative location for `benchmarking/README.md` and we might want to include benchmarking table output etc in that directory so I'm OK keeping it around as just a stub for now but you could also move it to e.g. `BENCHMARK.md` at the top-level.

> I've tried to think of ways to remove this redundancy but couldn't figure a way out that preserves the command name airl and removes airl from the named config.

I think this is fine, people can type the extra 5 keystrokes :)

ernestum commented:
`test_experiments` is now also crashing under MacOS, so I guess we need to fix that one too?

ernestum reviewed (approved):
LGTM
The windows pipeline will be fixed here: https://github.com/HumanCompatibleAI/imitation/pull/811
Can you merge this @adam?

AdamGleave commented:
Merged -- well done @taufeeque9 for pushing this over the finish line, and thanks @ernestum for the detailed review!
